s3stream/src/main/java/com/automq/stream/s3/index/LocalStreamRangeIndexCache.java

```diff
@@ -0,0 +1,312 @@
+                });
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    private <T> CompletableFuture<T> exec(Callable<T> r) {
+        return initCf.thenApply(v -> {
+            try {
+                return r.call();
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+        });
+    }
+
+    public void clear() {
+        writeLock.lock();
+        try {
+            streamRangeIndexMap.clear();
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+    public long nodeId() {
+        readLock.lock();
+        try {
+            return nodeId;
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    public CompletableFuture<Void> append(Map<Long, RangeIndex> rangeIndexMap) {
+        return exec(() -> {
+            writeLock.lock();
+            try {
+                for (Map.Entry<Long, RangeIndex> entry : rangeIndexMap.entrySet()) {
+                    long streamId = entry.getKey();
+                    RangeIndex rangeIndex = entry.getValue();
+                    streamRangeIndexMap.computeIfAbsent(streamId,
+                        k -> new SparseRangeIndex(COMPACT_NUM, SPARSE_PADDING)).append(rangeIndex);
+                }
+            } finally {
+                writeLock.unlock();
+            }
+            return null;
+        });
+    }
+
+    public CompletableFuture<Void> compact(Map<Long, RangeIndex> rangeIndexMap, Set<Long> compactedObjectIds) {
+        return exec(() -> {
+            writeLock.lock();
+            try {
+                for (Map.Entry<Long, RangeIndex> entry : rangeIndexMap.entrySet()) {
+                    long streamId = entry.getKey();
+                    RangeIndex rangeIndex = entry.getValue();
+                    streamRangeIndexMap.computeIfAbsent(streamId, k -> new SparseRangeIndex(COMPACT_NUM, SPARSE_PADDING))
+                        .compact(rangeIndex, compactedObjectIds);
+                }
+            } finally {
+                writeLock.unlock();
+            }
+            return null;
+        });
+    }
+
+    public CompletableFuture<Void> updateIndexFromRequest(CommitStreamSetObjectRequest request) {
+        Map<Long, RangeIndex> rangeIndexMap = new HashMap<>();
+        for (ObjectStreamRange range : request.getStreamRanges()) {
+            RangeIndex newRangeIndex = null;
+            if (request.getObjectId() != ObjectUtils.NOOP_OBJECT_ID) {
+                newRangeIndex = new RangeIndex(range.getStartOffset(), range.getEndOffset(), request.getObjectId());
+            }
+            rangeIndexMap.put(range.getStreamId(), newRangeIndex);
+        }
+        if (request.getCompactedObjectIds().isEmpty()) {
+            return append(rangeIndexMap);
+        }
+        return compact(rangeIndexMap, new HashSet<>(request.getCompactedObjectIds()));
+    }
+
+    public static ByteBuf toBuffer(Map<Long, SparseRangeIndex> streamRangeIndexMap) {
```

Comment:
Add magic to facilitate format evolution.

Reply:
solved in #1590